src/serialization/json_object.h

```diff
@@ -0,0 +1,393 @@
+//    materials provided with the distribution.
+// 
+// 3. Neither the name of the copyright holder nor the names of its contributors may be
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#pragma once
+
+#include "rapidjson/document.h"
+#include "cryptonote_core/cryptonote_basic.h"
+#include "rpc/message_data_structs.h"
+#include "cryptonote_protocol/cryptonote_protocol_defs.h"
+#include "common/sfinae_helpers.h"
+
+namespace cryptonote
+{
+
+namespace json
+{
+
+struct JSON_ERROR : public std::exception
+{
+  protected:
+    JSON_ERROR() { }
+    std::string m;
+
+  public:
+    virtual ~JSON_ERROR() { }
+
+    const char* what() const throw()
+    {
+      return m.c_str();
+    }
+};
+
+struct MISSING_KEY : public JSON_ERROR
+{
+  MISSING_KEY(const char* key)
+  {
+    m = std::string("Key \"") + key + "\" missing from object.";
+  }
+};
+
+struct WRONG_TYPE : public JSON_ERROR
+{
+  WRONG_TYPE(const char* type)
+  {
+    m = std::string("Json value has incorrect type, expected: ") + type;
+  }
+};
+
+struct BAD_INPUT : public JSON_ERROR
+{
+  BAD_INPUT()
+  {
+    m = "An item failed to convert from json object to native object";
+  }
+};
+
+struct PARSE_FAIL : public JSON_ERROR
+{
+  PARSE_FAIL()
+  {
+    m = "Failed to parse the json request";
+  }
+};
+
+// convenience macro
+#define OBJECT_HAS_MEMBER_OR_THROW(val, key) \
+  if (!val.HasMember(key)) \
+  { \
+    throw cryptonote::json::MISSING_KEY(key); \
+  }
+
+
+
+// POD to json value
+template <typename Type, typename = typename std::enable_if<sfinae::is_not_container<Type>::value, Type>::type>
+rapidjson::Value toJsonValue(rapidjson::Document& doc, const Type& pod);
+
+template <typename Type, typename = typename std::enable_if<sfinae::is_not_container<Type>::value, Type>::type>
+Type fromJsonValue(const rapidjson::Value& val);
+
+template <>
+rapidjson::Value toJsonValue(rapidjson::Document& doc, const std::string& i);
+
+template <>
+std::string fromJsonValue(const rapidjson::Value& i);
+
+rapidjson::Value toJsonValue(rapidjson::Document& doc, bool i);
+
+bool fromJsonValue(const rapidjson::Value& i);
```

Comment:
Was this supposed to be an explicit template specialization? This function is always invoked if no template parameter is specified, but never invoked if `fromJsonValue<bool>(val)` is specified. Weird.